src/libraries/System.Linq/src/System/Linq/ToCollection.cs

```diff
@@ -71,7 +71,25 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
             }
 #endif
 
-            return new List<TSource>(source);
+            return EnumerableToList(source);
+
+            [MethodImpl(MethodImplOptions.NoInlining)] // avoid large stack allocation impacting other paths
+            static List<TSource> EnumerableToList(IEnumerable<TSource> source)
```

Comment:
I'm hesitant to change this case. It's one thing when we have specialized knowledge of the source and we're able to actually influence that source's behavior. But this case is just the exact equivalent of `List<T>.ctor`, and I'm hard pressed to come up with an explanation for whey `new List<T>(arbitraryEnumerable)` would perform differently from `arbitraryEnumerable.ToList()`. If we believed this optimization to be truly important, why wouldn't it be done in List's ctor instead?

Reply:
> If we believed this optimization to be truly important, why wouldn't it be done in List's ctor instead?

`List<T>` is such a ubiquitous type that complicating the implementation could result in some serious bloat, especially if this means we get `ArrayPool<T>` generic instantiations for any `List<T>` where `T` is a struct. This could tilt the cost benefit analysis towards "not worth it" for `List<T>.ctor` and "worth it" for `ToList()`. _I don't know how trimming/code size is implemented though, so please ignore if wrong._

The other reason is that `System.Linq` is a different use-case to `SPC`; for hot paths (where common advice is to avoid LINQ) the slight penalty for low counts could be a trade-off not worth making, whereas in places LINQ is used it's probably preferable to create less garbage over raw speed.

These arguments is the best I can do, and I myself am not fully convinced. Either way please let me know your decision.